build/src/lib.rs

```diff
@@ -519,4 +703,76 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn build_with_zstd_transform() -> Result<()> {
+        let tempdir = tempdir::TempDir::new("builder_test")?;
+        let tempdir = tempdir.path();
+
+        let builder = BinaryBuilder::new(&["en"], tempdir)
+            .version("0.3.0")
+            .transform(
+                &|mut buffer, mut writer| {
+                    let mut data = Vec::new();
+                    buffer.read_to_end(&mut data)?;
+                    let data = smush::encode(
+                        data.as_slice(),
+                        smush::Codec::Zstd,
+                        smush::Quality::Maximum,
+                    )?;
+                    writer.write_all(&data)?;
+
+                    // XXX hangs forever, both with `smush` and `brotli` directly
+                    // let _ = brotli::BrotliCompress(&mut buffer, &mut sink, &Default::default())?;
+                    Ok(())
+                },
+                &|p: PathBuf| {
+                    let mut s = p.to_string_lossy().to_string();
+                    s.push_str(".zstd");
+                    Ok(PathBuf::from(s))
+                },
+            )
+            .build()?
+            .postprocess(
+                |mut buffer, mut writer| {
+                    dbg!("postprocessing");
+                    let mut tmp = Vec::new();
+                    buffer.read_to_end(&mut tmp)?;
+                    let data = smush::decode(tmp.as_slice(), smush::Codec::Zstd)?;
+                    writer.write_all(data.as_slice())?;
+                    Ok(())
+                    // XXX
+                    // Ok(brotli::BrotliDecompress(&mut buffer, &mut writer)?)
+                },
+                |p| {
+                    let path = p.to_string_lossy();
+                    assert!(path.ends_with(".zstd"));
+                    let end = path.len().saturating_sub(".zstd".len());
+                    assert_ne!(end, 0);
+                    path[..end].to_owned()
+                },
+            )?;
+
+        assert_eq!(
+            builder.outputs(),
+            &[
+                tempdir.join("en_tokenizer.bin"),
+                tempdir.join("en_rules.bin")
+            ]
+        );
+
+        let rules_path = tempdir
+            .join(Path::new(&rules_filename("en")))
+            .with_extension("bin");
```

Comment:
`.with_extension` is a noop here.
```suggestion
            .join(Path::new(&rules_filename("en")));
```